src/kernels/jupyter/finder/remoteKernelFinder.ts

```diff
@@ -67,102 +92,137 @@ export class RemoteKernelFinder implements IRemoteKernelFinder, IExtensionSingle
-                try {
-                    await Promise.race([kernelsFromCachePromise, kernelsWithoutCacheDeferred.promise]);
-                } catch (ex) {
-                    // If we failed to get without cache, then await on the cache promise as a fallback.
-                    if (kernelsWithoutCacheDeferred.rejected) {
-                        await kernelsFromCachePromise;
-                    } else {
-                        throw ex;
-                    }
-                }
-                // If we finish the cache first, and we don't have any items, in the cache, then load without cache.
-                if (Array.isArray(kernelsFromCache) && kernelsFromCache.length > 0) {
-                    kernels = kernelsFromCache;
-                } else {
-                    kernels = await kernelsWithoutCachePromise;
-                    updateCache = true;
-                }
-            } catch (ex) {
-                traceError(`Exception loading kernels: ${ex}`);
-            }
+            kernels = await kernelsWithoutCachePromise;
+        }
+
+        await this.writeToCache(kernels);
+        this._initializeResolve();
+    }
+
+    private async updateCache() {
+        let kernels: RemoteKernelConnectionMetadata[] = [];
+        this._cacheUpdateCancelTokenSource?.dispose();
+        const updateCacheCancellationToken = new CancellationTokenSource();
+        this._cacheUpdateCancelTokenSource = updateCacheCancellationToken;
+
+        try {
+            const kernelsWithoutCachePromise = (async () => {
+                const connInfo = await this.getRemoteConnectionInfo(updateCacheCancellationToken.token);
+                return connInfo ? this.listKernelsFromConnection(connInfo) : Promise.resolve([]);
+            })();
+
+            kernels = await kernelsWithoutCachePromise;
+        } catch (ex) {
+            traceWarning(`Could not fetch kernels from the ${this.kind} server, falling back to cache: ${ex}`);
+            // Since fetching the remote kernels failed, we fall back to the cache,
+            // at this point no need to display all of the kernel specs,
+            // Its possible the connection is dead, just display the live kernels we had.
+            // I.e. if user had a notebook connected to a remote kernel, then just display that live kernel.
+            kernels = await this.getFromCache(updateCacheCancellationToken.token);
+            kernels = kernels.filter((item) => item.kind === 'connectToLiveRemoteKernel');
         }
 
-        // Do not update the cache if we got kernels from the cache.
-        if (updateCache) {
-            await this.writeToCache(kernels);
+        if (updateCacheCancellationToken.token.isCancellationRequested) {
+            return;
         }
-        return kernels;
+
+        await this.writeToCache(kernels);
+
+        this._onDidChangeKernels.fire();
+    }
+
+    /**
+     * Used in test only
+     */
+    async _updateCacheForTest() {
```

Comment:
Can you not mark `loadCache` as public that way you dont need a method specifically for tests.

Reply:
No need to change, just a thought